onnxruntime/core/providers/cpu/nn/pool_functors.h

```diff
@@ -395,4 +395,309 @@ struct MaxPool3DTask {
+
+  T padding = std::numeric_limits<T>::lowest();
+  T* head = que;
+  T* tail = head;
+  int64_t cnt = 0; //num of elems in que
+  int64_t x_step = x_gap * dilation;
+  T* back = que + pool_size - 1;
+  const T* x_i = x - pad_front * x_gap;
+  const T* last_x_i = x_i;
+  const T* x_end_with_pads = x_end + pad_back * x_gap;
+
+  //init queue with first pool_size elems
+  for (int64_t i = 0; i < pool_size; i++, x_i += x_step) {
+    if (x_i < x) {
+      *tail = padding;
+      tail = tail == back ? que : tail + 1;
+      ++cnt;
+    } else {
+      if (cnt == 0) {
+        *tail = *x_i;
+        tail = tail == back ? que : tail + 1;
+        ++cnt;
+      } else {
+        T* last_le = nullptr;
+        if (head < tail) {
+          last_le = FindLastLargerEqual(head, tail, x_i);
+          cnt = last_le - head + 2;
+        } else {
+          if (*back >= *x_i) {
+            last_le = FindLastLargerEqual(que, tail, x_i);
+            cnt = back - head + last_le - que + 3;
+          } else {
+            last_le = FindLastLargerEqual(head, back + 1, x_i);
+            cnt = last_le - head + 2;
+          }
+        }
+        *++last_le = *x_i;
+        tail = last_le == back ? que : last_le + 1; 
+      }
+    }
+  }
+  
+  //fill max in current window
+  for (;; x_i += x_step, last_x_i += x_step, y += y_gap) {
+    //save max
+    *y = *head;
+    //dequeue
+    if (((last_x_i < x || last_x_i >= x_end) && Equal(*head, padding)) ||
+        (last_x_i >= x && last_x_i < x_end && Equal(*head, *last_x_i))) {
+      head = head == back ? que : head + 1;
+      --cnt;
+    }
+    if (x_i >= x_end_with_pads) {
+      break;
+    }
+    //enqueue
+    if (x_i < x || x_i >= x_end) {
+      *tail = padding;
+      tail = tail == back ? que : tail + 1;
+      ++cnt;
+    } else {
+      if (cnt == 0) {
+        *tail = *x_i;
+        tail = tail == back ? que : tail + 1;
+        ++cnt;
+      } else {
+        T* last_le = nullptr;
+        if (head < tail) {
+          last_le = FindLastLargerEqual(head, tail, x_i);
+          cnt = last_le - head + 2;
+        } else {
+          if (*back >= *x_i) {
+            last_le = FindLastLargerEqual(que, tail, x_i);
+            cnt = back - head + last_le - que + 3;
+          } else {
+            last_le = FindLastLargerEqual(head, back + 1, x_i);
+            cnt = last_le - head + 2;
+          }
+        }
+        *++last_le = *x_i;
+        tail = last_le == back ? que : last_le + 1; 
+      }
+    }
+  }
+}//MaxPoolOpt
+
+template <typename T>
+struct MaxPool1DTaskOpt {
+
+  const T* X_data;
+  T* Y_data;
+  int64_t x_h;
+  int64_t y_h;
+  int64_t pad_front;
+  int64_t pad_back;
+  int64_t dilation;
+  int64_t pool_size;
+
+  TensorOpCost Cost() {
+    auto loop_count = static_cast<double>(x_h);
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+#ifdef _OPENMP
+#pragma omp parallel for
```

Comment:
AFAIK- individual operator code shouldn't have OPENMP pragmas and they should use the "abstraction" ThreadPool class to parallelize shards of work - https://github.com/microsoft/onnxruntime/blob/master/docs/NotesOnThreading.md

Reply:
Removed.